src/rgw/rgw_common.cc

```diff
@@ -1467,3 +1478,78 @@ int rgw_parse_op_type_list(const string& str, uint32_t *perm)
   return parse_list_of_flags(op_type_mapping, str, perm);
 }
 
+static int match_internal(const char* pattern, const char* input)
+{
+  while(true) {
+    if (*pattern == '\0' && *input == '\0')
+      return 1;
+    if (*pattern == '\0' || *input == '\0')
+      return 0;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input != '\0')
+      return 1;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input == '\0')
+      return 0;
+    if (*pattern == *input || *pattern == '?') {
+      pattern++;
+      input++;
+      continue;
+    }
+    if (*pattern == '*') {
+      if (*(pattern + 1) == *input)
+        pattern++;
+      else
+        input++;
+      continue;
+    }
+    return 0;
+  }
+  return 0;
+}
+
```

Comment:
In this function, it would be better ouse boost::string_ref rather than using char* and a NUL terminator. That way we can pass in substrings easily without having to copy and NUL-terminate them.